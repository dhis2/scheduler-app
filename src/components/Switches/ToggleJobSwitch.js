import React from 'react'
import PropTypes from 'prop-types'
import { useDataMutation } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { Switch } from '@dhis2/ui'

/* istanbul ignore next */
const mutation = {
    resource: 'jobConfigurations',
    id: ({ id }) => id,
    type: 'update',
    partial: true,
    data: ({ enabled }) => ({ enabled }),
}

const ToggleJobSwitch = ({ id, checked, disabled, refetch }) => {
    const [toggleJob, { loading }] = useDataMutation(mutation)
    const enabled = !checked

    return (
        <Switch
            name={`toggle-job-${id}`}
            disabled={disabled || loading}
            checked={checked}
            onChange={() => {
                toggleJob({ id, enabled }).then(refetch)
            }}
            ariaLabel={i18n.t('Toggle job')}
        />
    )
}

const { bool, string, func } = PropTypes

ToggleJobSwitch.propTypes = {
    checked: bool.isRequired,
    disabled: bool.isRequired,
    id: string.isRequired,
    refetch: func.isRequired,
}

export default ToggleJobSwitch
