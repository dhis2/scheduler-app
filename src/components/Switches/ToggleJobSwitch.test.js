import React from 'react'
import { shallow, mount } from 'enzyme'
import { useDataMutation } from '@dhis2/app-runtime'
import ToggleJobSwitch from './ToggleJobSwitch'

jest.mock('@dhis2/app-runtime', () => ({
    useDataMutation: jest.fn(() => [() => {}, {}]),
}))

afterEach(() => {
    jest.resetAllMocks()
})

describe('<ToggleJobSwitch>', () => {
    it('renders without errors', () => {
        shallow(
            <ToggleJobSwitch
                id="1"
                checked={true}
                disabled={false}
                refetch={() => {}}
            />
        )
    })

    it('calls toggleJob and refetches when toggle is clicked', async () => {
        const checked = false
        const toggle = Promise.resolve()
        const toggleJobSpy = jest.fn(() => toggle)
        const refetchSpy = jest.fn(() => {})
        const props = {
            id: 'id',
            checked,
            disabled: false,
            refetch: refetchSpy,
        }

        useDataMutation.mockImplementation(() => [toggleJobSpy, {}])

        const wrapper = mount(<ToggleJobSwitch {...props} />)

        wrapper
            .find('input')
            .find({ name: 'toggle-job-id' })
            .simulate('change', { target: { checked: !checked } })

        await toggle

        expect(toggleJobSpy).toHaveBeenCalledWith({
            id: 'id',
            enabled: !checked,
        })
        expect(refetchSpy).toHaveBeenCalled()
    })
})
