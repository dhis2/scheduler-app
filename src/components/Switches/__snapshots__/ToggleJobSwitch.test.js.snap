// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleJobSwitch> renders correctly 1`] = `
<Switch
  checked={true}
  dataTest="dhis2-uicore-switch"
  name="toggle-job-1"
  onChange={[Function]}
/>
`;
